### PR TITLE
[WIP | SE] [Parse][Sema] Optional iteration

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
SE - https://github.com/apple/swift-evolution/pull/878

Beta version. 

**TODO**
* Figure out where to add executable tests.
* Add more comments, get rid of debug prints.
* Think about what can be done to avoid including `ConstriaintSystem.h` in `MiscDiagnostics.cpp` to reach necessary diagnostics APIs (`diagnoseUnwrap`).

### CODE REVIEW

File: include/swift/AST/Stmt.h [hunk 1]
```diff
@@ -808,25 +809,28 @@ class ForEachStmt : public LabeledStmt {
...
+              bool isOptional, Optional<bool> implicit = None)
     : LabeledStmt(StmtKind::ForEach, getDefaultImplicitFlag(implicit, ForLoc),
                   LabelInfo),
       ForLoc(ForLoc), Pat(nullptr), InLoc(InLoc), Sequence(Sequence),
-      WhereExpr(WhereExpr), Body(Body) {
+      WhereExpr(WhereExpr), Body(Body), isOptional(isOptional) {
     setPattern(Pat);
   }
-  
+
```
Comment (rudkx) [inline include/swift/AST/Stmt.h]: Please remove the whitespace-only diffs like this one (and several others).

Comment (AnthonyLatsis) [reply include/swift/AST/Stmt.h]: Those are from the last gardening commit (extra whitespaces I noticed *while* working on the proposal), I intendedly separated them. You want me to remove the commit?

Comment (rudkx) [reply include/swift/AST/Stmt.h]: Yes, you could replace it with just the typo fixes.

**Comment by DougGregor** [comment]
@swift-ci build toolchain

**Comment by AnthonyLatsis** [comment]
@rudkx What's the best place for tests that check the functioning of statements, i.e.

``` swift
let array: [Int]? = [1, 2, 3]
var counter = 0

for? _ in array { counter += 1 }
expectEqual(counter, 3) 
```

**Comment by rudkx** [comment]
@AnthonyLatsis People have been dropping small executable tests into `test/Interpreter` but I don't think that's ideal. The Development | Compiler forum might be a good place to discuss whether there is a better organization for executable tests.

**Comment by AnthonyLatsis** [comment]
Does that mean that we currently mainly rely on the standard library tests to verify that language constructs function correctly?

**Pull request closed without merging**